### PR TITLE
1, 2, 3, Kleisli!

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

**Comment by danbills** [comment]
I think you're subjecting yourself to a little more boilerplate than necessary. Instead of objects, aliases, and `convert`, I would just create vals in ast2wdlom which instantiate a CheckedAtoB

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/package.scala [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+
 import wdl.draft3.transforms.ast2wdlom._
+import wdl.draft3.transforms.wdlom2wom.FileElementToWomExecutable.FileElementAndInputsFile
 import wom.executable.Executable
 
 package object wdlom2wom {
   implicit val fromWorkflowDefinitionElementToWorkflowDefinition = WorkflowDefinitionElementToWomWorkflowDefinition
 
-  def womFromDraft3FileConverter(inputs: Option[String]): FromAtoB[File, Executable] = FromAtoB.viaX(draft3FileElementFromFile, FileElementToWomExecutable(inputs))
+  def womFromDraft3FileConverter(inputs: Option[String]): CheckedAtoB[File, Executable] = {
```
Comment (danbills) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/package.scala]: I think you will want this to be `def womFromDraft3FileConverter: CheckedAtoB[(File,  Option[String]), Executable]`, i.e. moving this argument into the Kleisli.  

If you go this route you might consider  creating a case class to give it better semantics

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/package.scala [hunk 2]
```diff
@@ -1,24 +1,44 @@
...
 
-  implicit val astFromAstNode = AstFromAstNode
-  implicit val draft3FileElementFromAstNode = FromAtoB.viaX(AstFromAstNode, Draft3FileElementFromAst)
-  implicit val draft3ImportElementFromAstNode = FromAtoB.viaX(AstFromAstNode, Draft3ImportElementFromAst)
-  implicit val draft3TaskDefinitionElementFromAstNode = FromAtoB.viaX(AstFromAstNode, Draft3TaskDefinitionElementFromAst)
-  implicit val draft3WorkflowDefinitionElementFromAstNode = FromAtoB.viaX(AstFromAstNode, Draft3WorkflowDefinitionElementFromAst)
+  type CheckedAtoB[A, B] = Kleisli[Checked, A, B]
+  object CheckedAtoB {
+    def apply[A, B](implicit runner: CheckedAtoB[A, B]): Kleisli[Checked, A, B] = runner
+    def apply[A, B](run: A => Checked[B]): Kleisli[Checked, A, B] = Kleisli(run)
```
Comment (danbills) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/package.scala]: ToL might as well declare `def apply[A, B](run: A => ErrorOr[B]): Kleisli[Checked, A, B] = Kleisli(run.map(_.toEither)`, save yourself conversions elsewhere

Comment (cjllanwarne) [reply wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/package.scala]: Bad news:
```
Error:(11, 9) double definition:
def apply[A, B](run: A => common.Checked[B]): cats.data.Kleisli[common.Checked,A,B] at line 10 and
def apply[A, B](run: A => common.validation.ErrorOr.ErrorOr[B]): cats.data.Kleisli[common.Checked,A,B] at line 11
have same type after erasure: (run: Function1)cats.data.Kleisli
    def apply[A, B](run: A => ErrorOr[B]): Kleisli[Checked, A, B] = Kleisli(runThenCheck(run))
```

Comment (cjllanwarne) [reply wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/ast2wdlom/package.scala]: so I can get around this by giving the apply's real names. A bit sad that I had to, but whatchagonnado...

File: common/src/main/scala/common/transforms/package.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package common
+
+import cats.data.Kleisli
+import common.validation.ErrorOr.ErrorOr
+
+package object transforms {
+  type CheckedAtoB[A, B] = Kleisli[Checked, A, B]
+  object CheckedAtoB {
+    def apply[A, B](implicit runner: CheckedAtoB[A, B]): Kleisli[Checked, A, B] = runner
```
Comment (mcovarr) [inline common/src/main/scala/common/transforms/package.scala]: why does this use `CheckedAtoB[A, B]` in one spot but not the other?

Comment (mcovarr) [reply common/src/main/scala/common/transforms/package.scala]: not in any of the return types actually

File: wom/src/main/scala/wom/transforms/WomExecutableMaker.scala [hunk 4]
```diff
@@ -3,9 +3,16 @@ package wom.transforms
...
 @typeclass
 trait WomExecutableMaker[A] {
-  def toWomExecutable(a: A, inputFile: Option[String] = None): Checked[Executable]
+  def toWomExecutable(inputs: ExecutableMakerInputs[A]): Checked[Executable]
+  def toWomExecutable(a: A, inputs: Option[String] = None): Checked[Executable]
 }
+
+object WomExecutableMaker {
+  final case class ExecutableMakerInputs[A](from: A, inputs: Option[String])
+}
```
Comment (mcovarr) [inline wom/src/main/scala/wom/transforms/WomExecutableMaker.scala]: \n

File: womtool/src/main/scala/womtool/graph/WomGraph.scala [hunk 5]
```diff
@@ -172,6 +172,7 @@ object WomGraph {
     // TODO: Remove this and the 'fakeInput' method with #2867
     val fakedInputs = JsObject(namespace.workflow.inputs map { i => i._1 -> fakeInput(i._2.womType) })
 
+
```
Comment (mcovarr) [inline womtool/src/main/scala/womtool/graph/WomGraph.scala]: 🔙

File: wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/parsing/package.scala [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+  // TODO: 2.11: the fromChecked version of this works fine in 2.12 but gets "Checked does not have a method map" in 2.11
+  // Rather than fighting that, I'm going to do this inefficient thing for now and let's switch back to fromChecked in April
+  implicit val checkedFileToWomExecutable = CheckedAtoB.fromErrorOr { (a: ExecutableMakerInputs[File]) =>
+    for {
+      fileElement <- checkedFileElementFromFile(a.from).toValidated
+      inputs = ExecutableMakerInputs(fileElement, a.inputs)
+      executable <- fileElementToWomExecutable(inputs).toValidated
+    } yield executable
+  }
```
Comment (mcovarr) [inline wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/parsing/package.scala]: not really sure what all is going on in this file

**Comment by cjllanwarne** [comment]
Pull approve is wrong, I have 3 approvals. Merging.

**Pull request merged**